charts/kyverno/templates/build-kyverno-trust/build-kyverno-trust.yaml

```diff
@@ -21,9 +21,15 @@ spec:
           image: {{ .Values.kyverno.buildKyvernoTrust.image.registry }}/{{ .Values.kyverno.buildKyvernoTrust.image.repository }}:{{ .Values.kyverno.buildKyvernoTrust.image.tag }}
           command:
             - '/bin/sh'
-          args:
             - '-c'
-            - 'cd /usr/local/sbin && sh build-trust.sh '
+            - |
+              echo "Running the Kyverno patch..."
+              cd /usr/local/sbin && sh build-trust.sh
+              echo "Waiting for Kyverno deployment to complete rollout..."
+              kubectl rollout status deployment kyverno-admission-controller -n kyverno --timeout=60s
+              echo "Kyverno deployment has successfully rolled out."
+              sleep 20
```

Comment:
Do you know whether the `kubectl rollout status` is succeeding, or timing out at 60 seconds?  If it's timing out, then rather than adding an additional 20 second sleep, why not just increase the timeout?  Or is it that the `kubectl rollout status` is succeeding, but kyverno still isn't ready?

I really dislike adding arbitrary sleeps to address timing issues, it would be better to loop checking for a specific condition, with a small wait in each loop.

Reply:
> Do you know whether the `kubectl rollout status` is succeeding, or timing out at 60 seconds? If it's timing out, then rather than adding an additional 20 second sleep, why not just increase the timeout? Or is it that the `kubectl rollout status` is succeeding, but kyverno still isn't ready?
> 
> I really dislike adding arbitrary sleeps to address timing issues, it would be better to loop checking for a specific condition, with a small wait in each loop.

After Kubectl rollout it is taking time to process the webhooks hence additional 20 second sleep. Comments are updated appropriately in the code.

By default Kyverno policy has 10 secs timeout. Based on the load the Kyverno community recommended to increase it by 20 secs. This was earlier documented but for this release we want to address this. Providing documentation link for reference : https://github.com/Cray-HPE/docs-csm/blob/release/1.5/operations/kubernetes/Kyverno.md#known-issues

Earlier this issue was rarely encountered but in CSM 1.6 release it is been encountered on multiple systems.

Reply:
I would implement this differently ... Logically, `build-trust` and `wait-for-webhook` are 2 different tasks and should be in 2 different Helm post-install hooks. For `wait-for-webhook` job, you can use `artifactory.algol60.net/csm-docker/stable/sdlc-ops/wait-for-it:1.0.0` image, which will wait for TCP response at given address and given port. here's a sample wait for PostgreSQL in cray-spire init container:
https://github.com/Cray-HPE/cray-spire/blob/master/charts/cray-spire/templates/server/deployment.yaml#L81-L87

Reply:
Thanks for the explanation and additional pointers.

Reply:
> I would implement this differently ... Logically, `build-trust` and `wait-for-webhook` are 2 different tasks and should be in 2 different Helm post-install hooks. For `wait-for-webhook` job, it's not sufficient to wait for deployment and then just add 20 seconds. There's no guarantee 20 seconds is enough. There's no post-wait check, so if there's an error or timeout, it will be ignored silently.
> 
> Instead, you can use `artifactory.algol60.net/csm-docker/stable/sdlc-ops/wait-for-it:1.0.0` image, which will wait for TCP response at given address and given port. Here's a sample wait for PostgreSQL in cray-spire init container: https://github.com/Cray-HPE/cray-spire/blob/master/charts/cray-spire/templates/server/deployment.yaml#L81-L87

Hi @mtupitsyn, We added both the logics(wait for & trust) in the same webhook. Going forward both of them are addressed when we upgrade to Kyverno 1.12.x version. If we use image logic for wait, then there will be additional CVE maintenance. For now we shall go-ahead with this logic.